### PR TITLE
Fix chained batch segfault

### LINKED ISSUE #620: Chained batch should be persistent
A segfault can happen when a chained batch is GC-ed in between `batch.write(callback)` and `BatchWriteWorker->DoExecute()`.

### PR DESCRIPTION
**Pull request by vweevers**
Closes #620.

I'll drop the last commit once CI is green.

### CODE REVIEW

File: binding.cc [hunk 1]
```diff
@@ -1773,21 +1773,30 @@ NAPI_METHOD(batch_clear) {
  */
 struct BatchWriteWorker final : public PriorityWorker {
   BatchWriteWorker (napi_env env,
+                    napi_value thisObj,
```
Comment (vweevers) [inline binding.cc]: Maybe I should I rename this to "context" or "batchObj". Other suggestions?

**Comment by vweevers** [comment]
AppVeyor failure is unrelated, restarting that job:

```
> electron@4.1.5 postinstall C:\projects\leveldown\node_modules\electron
> node install.js
C:\projects\leveldown\node_modules\electron\install.js:49
  throw err
  ^
Error: getaddrinfo ENOTFOUND github-production-release-asset-2e65be.s3.amazonaws.com github-production-release-asset-2e65be.s3.amazonaws.com:443
```

File: binding.cc (hunk 1 above)
Comment (peakji) [reply binding.cc]: `batchObj` sounds a little confusing alongside with `*batch`.

File: binding.cc [hunk 2]
```diff
@@ -1773,21 +1773,30 @@ NAPI_METHOD(batch_clear) {
...
       batch_(batch),
-      sync_(sync) {}
+      sync_(sync) {
+        // Prevent GC of batch object before we execute
+        NAPI_STATUS_THROWS(napi_create_reference(env_, thisObj, 1, &thisRef_));
+      }
 
-  ~BatchWriteWorker () {}
+  ~BatchWriteWorker () {
+    napi_delete_reference(env_, thisRef_);
```
[the remark this answers is not in the document]
Comment (vweevers) [reply binding.cc]: This destructor gets called via `FinalizeBatch()` during a GC cycle, which happens outside of a call stack where a JavaScript error _could_ be thrown. If that makes sense

File: binding.cc (hunk 1 above)
Comment (vweevers) [reply binding.cc]: Agree. What about "context"? That would match up with the name used in JS land:

https://github.com/Level/leveldown/blob/7312cbbe01c6191eaca8e8df7b99395211635585/chained-batch.js#L7

Comment (peakji) [reply binding.cc]: 👍

**Pull request merged**